### PR TITLE
Include more info in the meta/stats endpoint.

### PR DESCRIPTION
**Pull request by frankban**
The meta/stats endpoint returns aggregated downloads
count for the last day, week and month.
It also includes both the stats for the specific
entity revisions and the stats for all revisions of the
requested entity.

### CODE REVIEW

**Comment by frankban** [comment]
I'll update the spec accordingly later, if this proposal is accepted.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/608/

File: internal/v4/api_test.go [hunk 1]
```diff
@@ -1062,28 +1061,160 @@ func (s *APISuite) TestServeMetaRevisionInfo(c *gc.C) {
...
+	expectResponse: params.StatsResponse{
+		ArchiveDownloadCount:                      3 + 4 + 5,
+		ArchiveDownloadCountLastWeek:              5,
+		ArchiveDownloadCountLastMonth:             4 + 5,
+		ArchiveDownloadCountAllRevisions:          3 + 4 + 5,
+		ArchiveDownloadCountLastWeekAllRevisions:  5,
+		ArchiveDownloadCountLastMonthAllRevisions: 4 + 5,
+	},
+}, {
+	about: "bundle donwloads",
```
Comment (mhilton) [inline internal/v4/api_test.go]: *downloads

File: internal/v4/api_test.go [hunk 2]
```diff
@@ -1062,28 +1061,160 @@ func (s *APISuite) TestServeMetaRevisionInfo(c *gc.C) {
...
+		"utopic/wordpress-0": {},
+		"utopic/wordpress-1": {31: 7, 10: 1, 3: 2, 0: 1},
+		"utopic/wordpress-2": {6: 9, 0: 2},
+	},
+	expectResponse: params.StatsResponse{
+		ArchiveDownloadCountAllRevisions:          (7 + 1 + 2 + 1) + (9 + 2),
+		ArchiveDownloadCountLastDayAllRevisions:   1 + 2,
+		ArchiveDownloadCountLastWeekAllRevisions:  (2 + 1) + (9 + 2),
+		ArchiveDownloadCountLastMonthAllRevisions: (1 + 2 + 1) + (9 + 2),
+	},
```
Comment (mhilton) [inline internal/v4/api_test.go:1281]: Can you add a test for the same charm with different names. /trusty/mongodb and ~charmers/trusty/mongodb. I'm sure I read that these need to match somewhere.

**Comment by mhilton** [comment]
:+1: LGTM

File: internal/v4/api_test.go (hunk 1 above)
Comment (frankban) [inline internal/v4/api_test.go]: Done, thank you.

File: internal/v4/api_test.go (hunk 2 above)
Comment (frankban) [inline internal/v4/api_test.go:1281]: I don't think those are related, but a test documenting this is still useful, so I added that.
Thank you.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/609/

File: internal/v4/api.go [hunk 3]
```diff
@@ -405,21 +405,75 @@ func (h *Handler) metaTags(entity *mongodoc.Entity, id *charm.Reference, path st
...
+		return nil, errgo.Notef(err, "cannot get aggregated count for all revisions")
 	}
+
+	// Return the response.
 	return &params.StatsResponse{
-		// If a list is not requested as part of the charmstore.CounterRequest,
-		// one result is always returned: if the key is not found the count is
-		// set to 0.
-		ArchiveDownloadCount: results[0].Count,
+		ArchiveDownloadCount:                      counts.total,
```
Comment (rogpeppe) [inline internal/v4/api.go]: The names here are quite long. How about saving some bandwidth by encapsulating the results into a struct?
Even with the backward-compatibility field, it still saves 170 bytes per result, and the correspondence between the two sets of results is more obvious too.

```
type StatsResponse struct {
    ArchiveDownloadCount int64 // superceded by ArchiveDownload but maintained for backward compatibility.
    ArchiveDownload StatsCount
    ArchiveDownloadAllRevisions StatsCount
}

type StatsCount struct {
    Total int64                // Total count over all time.
    Day int64                 // Count over the last day.
    Week int64              // Count over the last week.
    Month int64             // Count over the last month.
}
```

File: internal/v4/api.go [hunk 4]
```diff
@@ -405,21 +405,75 @@ func (h *Handler) metaTags(entity *mongodoc.Entity, id *charm.Reference, path st
...
+// key.
+func (h *Handler) aggregatedStats(key []string) (aggregatedCounts, error) {
+	var counts aggregatedCounts
+
+	// Set prefix to true if the key is not complete. A complete key includes
+	// 5 items: the stats kind and the entity series, name, user and revision.
+	req := charmstore.CounterRequest{
+		Key:    key,
+		By:     charmstore.ByDay,
+		Prefix: len(key) < 5,
```
Comment (rogpeppe) [inline internal/v4/api.go]: How about making prefix an argument to this function? Then aggregatedStats becomes a useful general purpose function that can be used for any stats, regardless of their type, and doesn't need to know about the format of stats keys.

**Comment by rogpeppe** [comment]
Nice stuff. LGTM modulo a couple of suggestions.

File: internal/v4/api.go (hunk 4 above)
Comment (frankban) [inline internal/v4/api.go]: Good idea, done.

File: internal/v4/api.go (hunk 3 above)
Comment (frankban) [inline internal/v4/api.go]: Done.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/610/

**Comment by frankban** [comment]
Thank you both for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

File: internal/v4/api.go [hunk 5]
```diff
@@ -405,21 +405,78 @@ func (h *Handler) metaTags(entity *mongodoc.Entity, id *charm.Reference, path st
...
+			Day:   countsAllRevisions.lastDay,
+			Week:  countsAllRevisions.lastWeek,
+			Month: countsAllRevisions.lastMonth,
+		},
 	}, nil
 }
 
+// aggregatedStats returns the aggregated downloads counts for the given stats
+// key.
+func (h *Handler) aggregatedStats(key []string, prefix bool) (aggregatedCounts, error) {
```
Comment (rogpeppe) [inline internal/v4/api.go:442]: For a possible future change, this could return params.StatsCount, which would make things simpler above.

**Pull request merged**